### PR TITLE
Support autoscaling Cosmos DB databases and containers 

### LINKED ISSUE #6164: Support for cosmos db autopilot container
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Cosmos DB provides a new autopilot feature. This is currently in preview, but the new stable [2020-03](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2020-03-01/cosmos-db.json)  API support it.

### New or Affected Resource(s)

* azurerm_cosmosdb_sql_container

### References

* https://docs.microsoft.com/azure/cosmos-db/provision-throughput-autopilot
 * In the API it is called "provisionedThroughputSettings", see https://github.com/Azure/azure-rest-api-specs/pull/8609

### LINKED ISSUE #7315: Enable autoscaling for azurerm_cosmosdb_gremlin_graph
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

The current version of **azurerm_cosmosdb_gremlin_graph** does not support setting the throughput policy and autoscaling for Gremlin. See Microsoft resources:
a. https://azure.microsoft.com/en-gb/resources/templates/101-cosmosdb-gremlin-autoscale/
b. https://github.com/Azure/azure-quickstart-templates/tree/master/101-cosmosdb-gremlin-autoscale/

### New or Affected Resource(s)

* azurerm_cosmosdb_gremlin_graph

### Potential Terraform Configuration

```hcl
resource "azurerm_cosmosdb_gremlin_graph" "example" {
  name                = "tfex-cosmos-gremlin-graph"
  resource_group_name = azurerm_cosmosdb_account.example.resource_group_name
  account_name        = azurerm_cosmosdb_account.example.name
  database_name       = azurerm_cosmosdb_gremlin_database.example.name
  partition_key_path  = "/Example"
  throughput          = 400

  index_policy {
    automatic      = true
    indexing_mode  = "Consistent"
    included_paths = ["/*"]
    excluded_paths = ["/\"_etag\"/?"]
  }

  throughput_policy {
    autoscale = true
    max_throughput = 1000
  }

  conflict_resolution_policy {
    mode                     = "LastWriterWins"
    conflict_resolution_path = "/_ts"
  }

  unique_key {
    paths = ["/definition/id1", "/definition/id2"]
  }
}
```

### References
N/A

### PR DESCRIPTION
**Pull request by marc-sensenich**
Adds support for enabling autoscaling Cosmos DB databases and containers

Depends on fix for issue #7825 for testing
Depends on #7597 (merged in v2.20.0)
Closes #6164 
Closes #7315

### Example Usage

```tf
resource "azurerm_cosmosdb_sql_database" "autoscale" {
  account_name        = "myaccount"
  name                = "mydb"
  resource_group_name = "myrg"

  autoscale_settings {
    max_throughput = 4000
  }
}
```

### Test Status

Run with commit https://github.com/terraform-providers/terraform-provider-azurerm/pull/7773/commits/324d259c45a92912ac301d60927e992f7c2030a7

#### Passing
```
--- PASS: TestAccAzureRMCosmosDbCassandraKeyspace_basic (1330.26s)
--- PASS: TestAccAzureRMCosmosDbCassandraKeyspace_complete (1525.71s)
--- PASS: TestAccAzureRMCosmosDbCassandraKeyspace_update (1582.38s)
--- PASS: TestAccAzureRMCosmosDbCassandraKeyspace_autoscale (1778.28s)
--- PASS: TestAccAzureRMCosmosGremlinDatabase_requiresImport (1376.95s)
--- PASS: TestAccAzureRMCosmosGremlinDatabase_basic (1422.04s)
--- PASS: TestAccAzureRMCosmosGremlinDatabase_complete (1563.30s)
--- PASS: TestAccAzureRMCosmosGremlinDatabase_autoscale (1735.37s)
--- PASS: TestAccAzureRMCosmosDbGremlinGraph_requiresImport (1466.30s)
--- PASS: TestAccAzureRMCosmosDbGremlinGraph_customConflictResolutionPolicy (1498.55s)
--- PASS: TestAccAzureRMCosmosDbGremlinGraph_basic (1548.04s)
--- PASS: TestAccAzureRMCosmosDbGremlinGraph_indexPolicy (1610.39s)
--- PASS: TestAccAzureRMCosmosDbGremlinGraph_autoscale (1739.45s)
--- PASS: TestAccAzureRMCosmosDbGremlinGraph_update (1777.58s)
--- PASS: TestAccAzureRMCosmosDbMongoCollection_basic (1413.85s)
--- PASS: TestAccAzureRMCosmosDbMongoCollection_throughput (1624.28s)
--- PASS: TestAccAzureRMCosmosDbMongoCollection_autoscale (1744.72s)
--- PASS: TestAccAzureRMCosmosDbMongoDatabase_complete (1302.42s)
--- PASS: TestAccAzureRMCosmosDbMongoDatabase_basic (1307.51s)
--- PASS: TestAccAzureRMCosmosDbMongoDatabase_autoscale (1618.07s)
--- PASS: TestAccAzureRMCosmosDbSqlDatabase_basic (1350.69s)
--- PASS: TestAccAzureRMCosmosDbSqlDatabase_update (1500.18s)
--- PASS: TestAccAzureRMCosmosDbSqlDatabase_autoscale (1728.33s)
--- PASS: TestAccAzureRMCosmosDbSqlContainer_complete (1426.93s)
--- PASS: TestAccAzureRMCosmosDbSqlContainer_basic (1427.35s)
--- PASS: TestAccAzureRMCosmosDbSqlContainer_update (1572.95s)
--- PASS: TestAccAzureRMCosmosDbSqlContainer_autoscale (1734.78s)
--- PASS: TestAccAzureRMCosmosDbTable_basic (1380.36s)
--- PASS: TestAccAzureRMCosmosDbTable_autoscale (1788.62s)
--- PASS: TestAccAzureRMCosmosDbTable_update (1918.18s)
```

#### Failing

These failing tests are related to #7825 

```
--- FAIL: TestAccAzureRMCosmosDbMongoCollection_complete (1459.93s)
--- FAIL: TestAccAzureRMCosmosDbMongoCollection_withIndex (1502.19s)
--- FAIL: TestAccAzureRMCosmosDbMongoCollection_update (1639.81s)
```

### CODE REVIEW

File: azurerm/internal/services/cosmos/common/throughput.go
Comment (marc-sensenich) [inline azurerm/internal/services/cosmos/common/throughput.go:62]: From https://docs.microsoft.com/en-us/azure/cosmos-db/autoscale-faq#can-i-enable-autoscale-on-an-existing-database-or-a-container

> Yes. You can also switch between autoscale and standard (manual) provisioned throughput as needed. Currently, for all APIs, you can only use the Azure portal to do these operations.

Based on this and no clear documented API for toggling back and forth, I've decided to not allow this operation

**Comment by simongottschlag** [comment]
How is this PR going? 👍

**Comment by nadworny** [comment]
Does this PR also apply for azurerm_cosmosdb_mongo_database or only sql?

**Comment by marc-sensenich** [comment]
@simongottschlag this PR is complete and just awaiting review from an approver based on their priorities

@nadworny this PR covers all APIs that are supported by Terraform today

File: azurerm/helpers/validate/cosmos.go
Comment (WodansSon) [inline azurerm/helpers/validate/cosmos.go:49]: Can we switch all of the `value` variables to be `v` as well?

```suggestion
	if v < 4000 {
```

**Comment by marc-sensenich** [comment]
@WodansSon I've made the requested changes in https://github.com/terraform-providers/terraform-provider-azurerm/pull/7773/commits/7684202309a23401a1219ad09a645fcbe2153f46.

**Comment by marc-sensenich** [comment]
@WodansSon or @katbyte are there any further changes you'd like to see in this PR?

**Comment by joshatcaper** [comment]
Hey @tombuildsstuff it looks like this is tagged for `v2.26.0` and that seems to be due tomorrow. Is this going to make the merge window for that? I have a lot of use cases that could benefit from this and I'm really hoping we can start making use of this.

**Comment by marc-sensenich** [comment]
@tombuildsstuff, @WodansSon, and @katbyte are there any blockers to getting this into v2.27.0. I've addressed the requests by @WodansSon in https://github.com/terraform-providers/terraform-provider-azurerm/pull/7773/commits/beaf2f371680af313240736caf6560d507cef6cd

File: azurerm/internal/services/cosmos/common/throughput.go
Comment (manicminer) [reply azurerm/internal/services/cosmos/common/throughput.go:62]: There are some documented endpoints for migrating between autoscale and manual throughput (e.g. https://docs.microsoft.com/en-us/rest/api/cosmos-db-resource-provider/2020-04-01/tableresources/migratetabletoautoscale) however I couldn't find SDK support, so this seems fine :+1:

**Review by manicminer — CHANGES_REQUESTED** [review]
Hi @marc-sensenich, thanks for this contribution. This is looking great, following @WodansSon's review I have a few small suggestions and a query about ignoring/overwriting the manual throughput value returned by the API. Once these are looked at, I think we're good to merge and get into the next release 🚀

File: azurerm/internal/services/cosmos/common/throughput.go
Comment (marc-sensenich) [reply azurerm/internal/services/cosmos/common/throughput.go:57]: Let me investigate this one again

File: azurerm/internal/services/cosmos/common/autoscale_settings.go
Comment (marc-sensenich) [reply azurerm/internal/services/cosmos/common/autoscale_settings.go:25]: Let me investigate this one again

File: azurerm/internal/services/cosmos/common/throughput.go
Comment (marc-sensenich) [reply azurerm/internal/services/cosmos/common/throughput.go:57]: This masked that the API returns both the value for manual and autoscale throughput, I'm working on a code path that doesn't mask this behavior

**Comment by marc-sensenich** [comment]
@manicminer I've addressed your most recent comments. I'll run the full test-suite for Cosmos DB this evening and update this PR with the results

**Comment by marc-sensenich** [comment]
@manicminer Cosmos DB database and collection acceptance tests have been run from https://github.com/terraform-providers/terraform-provider-azurerm/commit/324d259c45a92912ac301d60927e992f7c2030a7 and are passing as noted in the initial PR message

**Comment by manicminer** [comment]
Test results:

Autoscaling:
Screenshot 2020-09-14 at 09 06 58

All (3 unrelated failures):
Screenshot 2020-09-14 at 09 08 12

**Comment by jonnekleijer** [comment]
@marc-sensenich 

Hi Marc, thanks for doing this great work! Does this mean we can already use this functionality in Azure? I was looking at the [documentation](https://www.terraform.io/docs/providers/azurerm/r/cosmosdb_sql_database.html) and there the autoscale option was not (yet) added.

**Comment by marc-sensenich** [comment]
@jonnekleijer this functionality will be available in the AzureRM provider in the to be released 2.28.0 version

**Pull request merged**
